domain/domain.go

```diff
@@ -633,37 +659,34 @@ func (do *Domain) mdlCheckLoop() {
 				// Schema doesn't change, and no job to check in the last run.
 				continue
 			}
-			// Get job to check.
-			rows, _, err := exec.ExecRestrictedSQL(kv.WithInternalSourceType(context.Background(), kv.InternalTxnTelemetry), nil, "select job_id, version, table_ids from mysql.tidb_mdl_info")
-			if err != nil {
-				continue
-			}
-			if len(rows) == 0 {
+
+			do.mdlCheckTableInfo.mu.Lock()
+
+			jobNeedToCheckCnt := len(do.mdlCheckTableInfo.jobsIdsMap)
+			if jobNeedToCheckCnt == 0 {
 				jobNeedToSync = false
+				do.mdlCheckTableInfo.mu.Unlock()
 				continue
 			}
 			jobNeedToSync = true
 
-			jobsVerMap := make(map[int64]int64, len(rows))
-			jobsIdsMap := make(map[int64]string, len(rows))
-			for i := 0; i < len(rows); i++ {
-				jobsVerMap[rows[i].GetInt64(0)] = rows[i].GetInt64(1)
-				jobsIdsMap[rows[i].GetInt64(0)] = rows[i].GetString(2)
-			}
-
 			sm := do.InfoSyncer().GetSessionManager()
 			if sm == nil {
 				logutil.BgLogger().Info("session manager is nil")
 			} else {
-				sm.CheckOldRunningTxn(jobsVerMap, jobsIdsMap)
+				sm.CheckOldRunningTxn(do.mdlCheckTableInfo.jobsVerMap, do.mdlCheckTableInfo.jobsIdsMap)
+			}
+
+			if len(do.mdlCheckTableInfo.jobsVerMap) == jobNeedToCheckCnt {
+				jobNeedToSync = false
 			}
 
-			// Try to gc jobsVerMap.
-			if len(jobsVerMap) > 1000 {
-				jobsVerMap = make(map[int64]int64, 1000)
+			// Try to gc jobCache.
+			if len(jobCache) > 1000 {
```

Comment:
GC will occur when item number > 1000, when. job Cache > 1000, it will start gc too frequently?